File: backend/app/schemas/driver.py
from datetime import date

from pydantic import BaseModel


class DriverBrief(BaseModel):
    id: int
    driver_ref: str
    number: int | None = None
    code: str | None = None
    first_name: str
    last_name: str
    nationality: str
    url: str | None = None
    photo_url: str | None = None

    class Config:
        from_attributes = True


class SeasonStatsOut(BaseModel):
    races: int = 0
    wins: int = 0
    podiums: int = 0
    poles: int = 0
    points: float = 0
    final_position: int | None = None
    constructor_name: str | None = None


class DriverDetail(BaseModel):
    id: int
    driver_ref: str
    number: int | None = None
    code: str | None = None
    first_name: str
    last_name: str
    date_of_birth: date | None = None
    nationality: str
    url: str | None = None
    photo_url: str | None = None
    season_stats: SeasonStatsOut | None = None

    class Config:
        from_attributes = True


class DriverRaceResultOut(BaseModel):
    round: int
    race_name: str
    circuit_name: str
    date: date
    grid_position: int | None = None
    position: int | None = None
    position_text: str | None = None
    points: float = 0
    laps_completed: int | None = None
    finish_time: str | None = None
    fastest_lap_time: str | None = None
    fastest_lap_speed: float | None = None
    fastest_lap_rank: int | None = None
    status: str | None = None
    qualifying_position: int | None = None


class DriverSeasonStats(BaseModel):
    driver: DriverBrief
    constructor_name: str
    total_points: float
    best_finish: int | None = None
    wins: int
    podiums: int
    poles: int
    fastest_laps: int
    races_entered: int
    final_position: int | None = None

    class Config:
        from_attributes = True

File: frontend/src/types/driver.ts
export interface DriverBrief {
  id: number;
  driver_ref: string;
  number?: number;
  code?: string;
  first_name: string;
  last_name: string;
  nationality?: string;
  url?: string;
  photo_url?: string;
}

export interface DriverSeasonStats {
  year: number;
  constructor_ref: string;
  constructor_name: string;
  races: number;
  wins: number;
  podiums: number;
  poles: number;
  fastest_laps: number;
  points: number;
  final_position?: number;
}

export interface DriverDetail extends DriverBrief {
  season_stats?: DriverSeasonStats;
}

export interface DriverRaceResult {
  round: number;
  race_name: string;
  circuit_name: string;
  date: string;
  grid_position: number | null;
  position: number | null;
  position_text: string | null;
  points: number;
  laps_completed: number | null;
  finish_time: string | null;
  fastest_lap_time: string | null;
  fastest_lap_speed: number | null;
  fastest_lap_rank: number | null;
  status: string | null;
  qualifying_position: number | null;
}
